--- a/net/minecraft/world/entity/ai/gossip/GossipContainer.java
+++ b/net/minecraft/world/entity/ai/gossip/GossipContainer.java
@@ -38,6 +_,7 @@
     private GossipContainer(List<GossipContainer.GossipEntry> entries) {
         entries.forEach(entry -> this.getOrCreate(entry.target).entries.put(entry.type, entry.value));
     }
+    private final UUID globalGossipsId = UUID.fromString("38400000-8cf0-11bd-b23e-10b96e4ef00d");
 
     @VisibleForDebug
     public Map<UUID, Object2IntMap<GossipType>> getGossipEntries() {
@@ -104,6 +_,15 @@
             }
         });
     }
+    public int getGlobalReputation()
+    {
+        GossipContainer.EntityGossips entityGossips = this.gossips.get(globalGossipsId);
+        Predicate<GossipType> gossipTypeFilter = (reputationtype) -> {
+            return true;
+        };
+        return entityGossips != null ? entityGossips.weightedValue(gossipTypeFilter) : 0;
+    }
+
 
     public int getReputation(UUID identifier, Predicate<GossipType> gossip) {
         GossipContainer.EntityGossips entityGossips = this.gossips.get(identifier);
@@ -112,6 +_,12 @@
 
     public long getCountForType(GossipType gossipType, DoublePredicate gossipPredicate) {
         return this.gossips.values().stream().filter(gossips -> gossipPredicate.test(gossips.entries.getOrDefault(gossipType, 0) * gossipType.weight)).count();
+	}
+	
+    public void addWithGlobal(UUID target, GossipType type, int value)
+    {
+        this.add(globalGossipsId, type, value);
+        this.add(target, type, value);
     }
 
     public void add(UUID identifier, GossipType gossipType, int gossipValue) {
@@ -220,6 +_,44 @@
         public void remove(GossipType gossipType) {
             this.entries.removeInt(gossipType);
         }
+
+        // Paper start - Add villager reputation API
+        private static final GossipType[] TYPES = GossipType.values();
+
+        public com.destroystokyo.paper.entity.villager.Reputation getPaperReputation() {
+            Map<com.destroystokyo.paper.entity.villager.ReputationType, Integer> map = new java.util.EnumMap<>(com.destroystokyo.paper.entity.villager.ReputationType.class);
+            for (Object2IntMap.Entry<GossipType> type : this.entries.object2IntEntrySet()) {
+                map.put(toApi(type.getKey()), type.getIntValue());
+            }
+
+            return new com.destroystokyo.paper.entity.villager.Reputation(map);
+        }
+
+        public void assignFromPaperReputation(com.destroystokyo.paper.entity.villager.Reputation rep) {
+            for (GossipType type : TYPES) {
+                com.destroystokyo.paper.entity.villager.ReputationType api = toApi(type);
+
+                if (rep.hasReputationSet(api)) {
+                    int reputation = rep.getReputation(api);
+                    if (reputation == 0) {
+                        this.entries.removeInt(type);
+                    } else {
+                        this.entries.put(type, reputation);
+                    }
+                }
+            }
+        }
+
+        private static com.destroystokyo.paper.entity.villager.ReputationType toApi(GossipType type) {
+            return switch (type) {
+                case MAJOR_NEGATIVE -> com.destroystokyo.paper.entity.villager.ReputationType.MAJOR_NEGATIVE;
+                case MINOR_NEGATIVE -> com.destroystokyo.paper.entity.villager.ReputationType.MINOR_NEGATIVE;
+                case MINOR_POSITIVE -> com.destroystokyo.paper.entity.villager.ReputationType.MINOR_POSITIVE;
+                case MAJOR_POSITIVE -> com.destroystokyo.paper.entity.villager.ReputationType.MAJOR_POSITIVE;
+                case TRADING -> com.destroystokyo.paper.entity.villager.ReputationType.TRADING;
+            };
+        }
+        // Paper end - Add villager reputation API
     }
 
     record GossipEntry(UUID target, GossipType type, int value) {
